cpp/include/cuspatial/detail/cartesian_product_iterator.cuh

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required point_b_y applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/pair.h>
+#include <utility>
+#include "utility/size_from_offsets.cuh"
+
+namespace cuspatial {
+namespace detail {
+
+struct cartesian_product_group {
+  int32_t idx;
+  int32_t size;
+  int32_t offset;
+};
+
+struct cartesian_product_group_index {
+  cartesian_product_group group_a;
+  cartesian_product_group group_b;
+  int32_t element_a_idx;
+  int32_t element_b_idx;
+};
+
+template <typename OffsetIteratorA,
+          typename OffsetIteratorB,
+          typename SizeIteratorLHS,
+          typename SizeIteratorRHS,
+          typename LookupIteratorLHS,
+          typename LookupIteratorRHS>
+struct cartesian_product_functor {
+  int32_t const num_elements_b;
+  SizeIteratorLHS const group_sizes_a;
+  SizeIteratorRHS const group_sizes_b;
+  OffsetIteratorA const group_offsets_a;
+  OffsetIteratorB const group_offsets_b;
+  LookupIteratorLHS const group_lookup_a;
+  LookupIteratorRHS const group_lookup_b;
+
+  __device__ cartesian_product_group_index operator()(int64_t const idx) const
```

Comment:
Why is `idx` a `int64_t`? It immediately gets converted to `int32_t` inside this function.

Reply:
`2^31 - 1`, but I get the point. :)